scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala

```diff
@@ -0,0 +1,104 @@
+package com.twitter.scalding.spark_backend
+
+import com.twitter.scalding.{ExecutionCounters, StatKey}
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.util.AccumulatorV2
+
+import scala.collection.mutable.{Map => MutableMap}
+
+/**
+ * Spark Counters is a singleton responsible for managing counters in the spark backend
+ *
+ * Inspiration has been taken from
+ * https://stackoverflow.com/questions/66438570/create-accumulator-on-executor-dynamically
+ */
+object SparkCounters {
+
+  var accumulator: SparkCountersInternal = null
+
+  /**
+   * This should be called once prior to starting execution on the spark backend. The spark context must
+   * already be active in the JVM.
+   */
+  def register(): Unit = {
+    val sparkContext = SparkSession.getActiveSession
+      .getOrElse(throw new IllegalStateException("SparkSession is not active!"))
+      .sparkContext
+    val accum = new SparkCountersInternal
+    sparkContext.register(accum, "[scalding/spark-backend] accumulator")
+    accumulator = accum
+  }
+
+  /**
+   * Adds a long value to a (counterGroup, counterName). Should be used inside of executors for side-effect
+   * free counter implementation
+   */
+  def addToCounter(counterGroup: String, counterName: String, diffVal: Long): Unit =
+    getAccumulator.add(counterGroup, counterName, diffVal)
+
+  /**
+   * we return an ExecutionCounters interface.
+   *
+   * Values returned from keys/get will not be stable until spark execution is complete
+   */
+  def lazyEvaluateAsExecutionCounters(): ExecutionCounters =
+    new ExecutionCounters {
+      def keys =
+        getAccumulator.value.foldLeft(Set[StatKey]())((ssk, kvPair) =>
+          ssk + StatKey(kvPair._1._1, kvPair._1._2)
+        )
+      def get(key: StatKey) =
+        getAccumulator.value.get((key.group, key.counter))
+    }
+
+  private def getAccumulator(): SparkCountersInternal = {
+    if (accumulator == null) {
+      throw new IllegalStateException("SparkCounters.register has not been called!")
+    }
+    accumulator
+  }
+
+}
+
+sealed class SparkCountersInternal(
+    val mapState: MutableMap[(String, String), Long] = MutableMap[(String, String), Long]()
+) extends AccumulatorV2[(String, String, Long), MutableMap[(String, String), Long]]
```

Comment:
if you make this `AccumulatorV2[((String, String), Long), ...]` can you avoid reboxing the counters?

Reply:
done